### PR TITLE
Efficient dry run with binary search tree

### PR DESCRIPTION
**Pull request by johugger**
The purpose of the 'dry run' is to determine a block shape (shape of tensors), which a loaded neural network can process on a GPU without knowing anything about the network. Therefore, the first call of the 'dry run' method creates a binary search tree of potential 'valid' block shapes and then uses this tree for an efficient lookup of the optimal shape. Every subsequent call (at a later point during runtime) then uses this tree to infer an optimal shape.

### CODE REVIEW

File: tiktorch/utils.py [hunk 1]
```diff
@@ -112,3 +113,150 @@ def base_shape(self):
...
+                    return recursive_search(x.left, last_valid_node)
+                else:
+                    return last_valid_node
+
+        return recursive_search(self.root)
+
+
+def test_binary_search_tree():
+    import torch.nn as nn
+    inputs = [[6, 32], [6, 112], [6, 272], [9, 32], [9, 192],
```
Comment (m-novikov) [inline tiktorch/utils.py]: if we would have sorted input here:
```
input.sort(key=lambda o: o[0] * o[1])
```
Then we will get an unbalanced tree with only left nodes(or right depending on sort order).
Maybe we can use `bisect` module here from stdlib for binary search?

Comment (m-novikov) [reply tiktorch/utils.py]: And if we can use simple binary search in a list, we don't actually need to build a tree,
sorting shapes has negligible cost
```python
TH = 1000

def allocate(size):
    if size > 1000:
        raise Exception('oom')

def find(sizes, lo=0, hi=None):
    # precond: list is sorted
    if not sizes:
        raise ValueError()

    if hi is None:
        hi = len(sizes) - 1

    if lo > hi:
        raise ValueError()

    mid = lo + (hi - lo) // 2

    if hi == lo:
        try:
            allocate(sizes[hi])
            return sizes[hi]
        except Exception:
            raise Exception('No valid shapes found')

    try:
        allocate(sizes[mid])
    except Exception:
        return find(sizes, lo=lo, hi=mid)
    else:
        return find(sizes, lo=mid, hi=hi)


def test_binary():
    sizes = [100, 200, 300, 500, 900, 1001, 2000, 3000]
    assert 900 == find(sizes)

    sizes = [100]

    assert 100 == find(sizes)

    sizes = [300]

    assert 300 == find(sizes)
```

File: tests/test_server.py [hunk 2]
```diff
@@ -82,3 +82,13 @@ def test_forward_pass(datadir, srv, client, nn_sample):
     res = client.forward(NDArrayBatch([NDArray(input_arr)]))
     res_numpy = res.as_numpy()
     np.testing.assert_array_almost_equal(res_numpy[0], out_arr)
+
+
+def test_client_dry_run(srv, client, nn_sample):
+    client.load_model(nn_sample.config, nn_sample.model, nn_sample.state, b'')
+    valid_shape = client.dry_run({
+        'train': False,
```
Comment (FynnBe) [inline tests/test_server.py]: We should test the dry run for training as well

Comment (johugger) [reply tests/test_server.py]: Yes! But it's not implemented yet.

Comment (johugger) [reply tests/test_server.py]: Done

File: tiktorch/utils.py (hunk 1 above)
Comment (johugger) [reply tiktorch/utils.py]: Thank you. Yes, I agree and it also simplifies the code. I have removed the binary search tree in favour of a sorted list + binary search.

**Pull request merged**